broaden dummy_eq capabilities

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

fixes #15240 while retaining `Sum(x, (x, 1, 2)) != Sum(y, (y, 1, 2))`
fixes #14324

#### Brief description of what is fixed or changed

Subs and Lambda compute hashable content
so expressions that are trivially different compare the same.
The same has not been done for ExprWithLimits. To aid in
their comparison (and in the general representation of objects
with structural dummy symbols) the `as_dummy` method now
can return non-clashing Symbol instances instead of Dummy.

Any object that defines a `bound_symbols` property will be
targeted by symbol-substitution when calling `as_dummy` on
an expression.


#### Other comments

The `as_dummy` function will now allow for simplification of expressions and canonical representation of objects that involve dummy (structural or otherwise) symbols:

```python
>>> eq=4*y*Sum(z, (z, n, k)) + 3*Sum(x**3 + x, (x, n, k)) + 1
>>> simplify(eq)
4*y*Sum(z, (z, n, k)) + 3*Sum(x**3 + x, (x, n, k)) + 1
>>> eq.as_dummy()
4*y*Sum(_0, (_0, n, k)) + 3*Sum(_0**3 + _0, (_0, n, k)) + 1
>>> simplify(_).subs(var('_0'), i)
Sum(u*(3*i**2 + 4*y + 3), (i, n, k)) + 1
```
Here, a dummy variable of integration is returned. It can be folded to a canonical form:
```python
>>> dsolve(x*f(x).diff(x) - f(x) - x*sin(f(x)/x), f(x), 
... hint='1st_homogeneous_coeff_subs_indep_div_dep_Integral',
... simplify=False)
Eq(log(f(x)), log(C1) + Integral(-(_u2 + 1/sin(1/_u2))/_u2**2, (_u2, x/f(x))))
>>> _.as_dummy()
Eq(log(f(x)), log(C1) + Integral(-(_0 + 1/sin(1/_0))/_0**2, (_0, x/f(x))))
```
The folded symbols are Symbol objects, not Dummy, so they should print well and are WYSIWYG when it comes to replacement. They will never clash with existing variables. If the ode above is written, with malice aforethought, in term of `_0` instead of `x`, then the result of folding is
```python
Eq(log(f(_u)), log(C1) + Integral(-(_1 + 1/sin(1/_1))/_1**2, (_1, _0/f(_0))))
```
#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * `dummy_eq` now compares expressions with canonical dummy variables
  * `canonical_variables` now returns non-clashing Symbols with simple names
* concrete
  * `bound_symbols` returns the bound symbols of the expr

<!-- END RELEASE NOTES -->

asmeurer commented on sympy/solvers/ode.py:
```diff
@@ -3478,13 +3479,13 @@ def ode_Bernoulli(eq, func, order, match):
 
     >>> pprint(dsolve(Eq(f(x).diff(x) + P(x)*f(x), Q(x)*f(x)), f(x),
-    ... hint='separable_Integral'))
+    ... hint='separable_Integral').fold())
```
Why did you add `fold` to these docstrings? The dummy variable name `y` is used here intentionally, to represent f(x).

smichr replied:
I did it because for debugging purposes it is useful to have dummies created in different areas of the code use different names. Then these are emitted in the final solution. Even if that wasn't done, the Dummy `_y` that is returned is not that attractive and makes testing fragile (because you have to test strings). Would you be ok with calling fold on the return value so u is always used? It is clear from the upper limit of the integral what that value will be (e.g. `f(x)` or `x**n*f(x)` in some places).

asmeurer commented:
This looks like a more reasonable solution, though I'm unclear why you have added `fold` in some places.

smichr committed:
broaden dummy_eq capabilities

Subs has been given a method of computing hashable content
so expressions that are trivially different compare the same.
The same has not been done for ExprWithLimits. To aid in
their comparison (and in the general representation of objects
with structural dummy symbols) the following modifications have
been made:

* core
  * `dummy_eq` now compares expressions with canonical dummy variables
  * `canonical_variables` now returns non-clashing Symbols
  * `dummy_eq` allows keyword `canonical` to replace structural dummy variables
     with canonical symbols (u, v, w, x, y, z, iota_0, iota_1, ...)
* concrete
  * dummy_variables returns the structural dummy variables of the expr

Any object that has structural dummies should define a `dummy_variables`
property that returns an iterable of its dummy variables.

smichr commented:
I see that pprinting dumy y doesn't show the underscore. I removed the modifications to ode and just used the fold in the tests.

asmeurer commented:
I would also suggestion using a more specific name like `fold_dummy`.

smichr commented:
> I would also suggest

I folded the behavior into the existing `as_dummy`.

smichr commented:
I think this is in good shape now.

smichr commented:
Not sure why I am getting a [sphinx error](https://travis-ci.org/sympy/sympy/jobs/436264679). Any ideas, @sylee957 ?

sylee957 commented:
I think matplotlib is introduced to sphinx in #15310 
Probably should be rebased

smichr committed:
add tests for Subs and Lambda

smichr commented:
@jksuom , @asmeurer do you have any thoughts on this? If not, I will commit in 3 days.

jksuom commented:
Would it be possible to extend the meaning of `canonical`. It might sometimes be helpful to be able to choose the symbols. Maybe something like `canonical='ijk'` for the first few symbols.

smichr commented:
> extend the meaning of `canonical`

Yes, I thought of that but initially rejected the idea since in the unnested context, the user can select their own variable as in `Sum(i, (i, a, b))`. In an unnested expression from something like dsolve that uses a dummy in the Integral form, the replacement of the symbol via as_dummy is 
```
foo.as_dummy(canonical='i')
```
vs
```
foo.as_dummy().subs(u, i)
```
In the nested context, the user might have a hard time figuring out which one will be assigned to which variable.

If routines that send back expressions containing dummy symbols would always use `return rv.as_dummy()` then the user would always be able to target the dummy variable through subs without the use of `as_dummy`.

But I am not opposed to the idea. What about the following semantics for canonical:
```
if canonical and type(canonical, string_type):
    use canonical + numbered_symbols('iota_')
elif canonical:
    use default
else:
    canonical = False
```
or changing args to `canonical=False, plain=False, sym='uvwxyz',num='iota_')` so the example above would be `foo.as_dummy(canonical=True, sym='i')  # or foo.as_dummy(1, 0, 'i')` or maybe the use always wants numbered alphas so they use `foo.as_dummy(1, gen='alpha_')`.

Let me know what you think.

jksuom commented:
> In the nested context, the user might have a hard time figuring out which one will be assigned to which variable.

This might be even programmatically challenging. One should use the same substitution for all top level variables and then another for the next level, etc.

> or changing args to canonical=False, plain=False, sym='uvwxyz',num='iota_')

I think that the api should be simple. There is probably seldom need to go beyond level three. If the numbered symbols would ever be needed, they could probably be hard-coded.

We could likely also do without `canonical`. It could be the default if nothing else is given.

I also think that assumptions on summation indices or integration variables should be ignored. They are restricted to certain domains and that should override any assumptions. So the keyword `plain` should not be necessary.

Hence it seems to me that it would suffice to give the `sym` argument alone, with `canonical` understood if no symbols are given. However, there is one change that I would like to make. There are two types of variables, discrete and continuous. It would be nice if different names could be used, for example, `ijkl` and `stuv`. They could probably be positional arguments so it would not be necessary to always add keywords.

Finally, I wonder if the name 'dummy' could be confusing as we also have `Dummy` symbols. The non-free variables in question are often called 'bound' variables. Maybe that could be used instead of 'dummy'.

smichr commented:
So are you thinking that we have `def as_dummy(sym=None):` as the signature with sym=None giving the current behavior in master, `sym='abc'` means use `symbols('abc')` then numbered `iotas`. 

I agree about the assumptions, but I was trying to retain compatibility with the current behavior. I'll see if that can be removed, however.

I am not sure what you mean about `ijkl` and `stuv`. Are those values that you would feed to `sym`? Can you give me an example?

Regarding `dummy`. Since we already have `as_dummy` I am assuming that you aren't talking about that; do you mean the `dummy_variables` property? If so, that could be changed to `bound_symbols` (in contrast to `free_symbols`.

asmeurer commented:
How about `canonical=<iterator of names>` (if the iterator runs out it falls back to the default ones).

smichr committed:
dummy_variables -> bound_symbols; syms and default

jksuom commented:
> Do you mean Sum(i, (i, 1, 10)) vs Integral(x, (x, 1, 10)).

Yes. These two types of objects cannot be added and combined into a single Sum or Integral. (Besides, there is an unwritten convention on the use of letters at the end of the alphabet for continuous variables and those in the middle for indices and integral values. That is only a stylistic convention but it makes it easier to remember the role of the symbols.)

> How about canonical='iterator of names' (if the iterator runs out it falls back to the default ones).

This is what I thought, but with two iterators instead of a common one for both indices and integration variables. It would not be necessary to include the name `canonical`. The two iterators could be positional parameters (maybe with defaults `None`). They could also be empty, in which case one would only use the hard coded lists (one for each type of symbols).

smichr commented:
Is there any reason we don't just enforce a uniform convention for dummy symbols at instantiation? e.g. `Sum(x, (x, 1, 2))` would automatically become `Sum(i, (i, 1, 2))` while `Sum(x + i, (x, 1, 2))` would become `Sum(i_0 + i, (i_0, 1, 2))`?

Regarding the use of different symbols for different types of objects: pehaps a dictionary input can be added to `as_dummy` that provides indices to use for the objects given as keys, e.g. `as_dummy(ix={Sum: 'ijk', Integral: 'xyz'})` and then those would be the ones to start with for that type of object (and the the numbered defaults).

smichr committed:
narrow as_dummy capabilities

smichr commented:
Keeping it simple: `as_dummy` is modified to allow `dummy_eq` to work. It can also be used to help reduce objects with bound symbols into a common symbol set wrt bound symbols. But that is not the main focus and symbol twiddling is left the the user.

smichr commented:
- [ ] remove bound_symbols from Derivative to fix error
